nexus/db-queries/src/db/datastore/affinity.rs

```diff
@@ -0,0 +1,2701 @@
+                let err = err.clone();
+                use db::schema::affinity_group::dsl as group_dsl;
+                use db::schema::affinity_group_instance_membership::dsl as membership_dsl;
+                use db::schema::instance::dsl as instance_dsl;
+
+                async move {
+                    // Check that the group exists
+                    group_dsl::affinity_group
+                        .filter(group_dsl::time_deleted.is_null())
+                        .filter(group_dsl::id.eq(authz_affinity_group.id()))
+                        .select(group_dsl::id)
+                        .first_async::<uuid::Uuid>(&conn)
+                        .await
+                        .map_err(|e| {
+                            err.bail_retryable_or_else(e, |e| {
+                                public_error_from_diesel(
+                                    e,
+                                    ErrorHandler::NotFoundByResource(
+                                        authz_affinity_group,
+                                    ),
+                                )
+                            })
+                        })?;
+
+                    // Check that the instance exists, and has no VMM
+                    //
+                    // NOTE: I'd prefer to use the "LookupPath" infrastructure
+                    // to look up the path, but that API does not give the
+                    // option to use the transaction's database connection.
+                    //
+                    // Looking up the instance on a different database
+                    // connection than the transaction risks several concurrency
+                    // issues, so we do the lookup manually.
+                    let instance_state = instance_dsl::instance
+                        .filter(instance_dsl::time_deleted.is_null())
+                        .filter(instance_dsl::id.eq(instance_id))
+                        .select(instance_dsl::state)
+                        .first_async(&conn)
+                        .await
+                        .map_err(|e| {
+                            err.bail_retryable_or_else(e, |e| {
+                                public_error_from_diesel(
+                                    e,
+                                    ErrorHandler::NotFoundByLookup(
+                                        ResourceType::Instance,
+                                        LookupType::ById(instance_id)
+                                    ),
+                                )
+                            })
+                        })?;
+
+                    // NOTE: It may be possible to add non-stopped instances to
+                    // affinity groups, depending on where they have already
+                    // been placed. However, only operating on "stopped"
+                    // instances is much easier to work with, as it does not
+                    // require any understanding of the group policy.
```

Comment:
What about `Creating` and `Failed` instances? Instances in those states also don't have a running VMM...although it seems unlikely that anyone would try to add them to affinity groups while they're in those states...

Upon consideration, it's probably better not to allow that, since both of those states might result in an automatic attempt to start the instance (either due to creating it with `start: true` or reincarnation) while `NoVmm` instances will only be started if someone decides to start them. So maybe it reduces the risk of racing with a start attempt if we're more conservative here.

Reply:
My gut tells me it's not important to worry about existing instances with VMMs, at least not now.

Reply:
Ugh, I think that given my prior comments about "adding a sled reservation is not atomic with VMM record creation", this code is actually wrong.

I think I need to be checking "does a sled resource record exist for this instance", rather than the VMM table's state.

If "yes", for any reason, I think we need to deny this request. That's the crux of this -- if a sled reservation record exists, the instance has a "right" to exist on a sled, and we would need to re-evaluate that against affinity rules if it's the case.

Reply:
Oh, yeah good catch.

Reply:
Patched this in 1ad01010bcd8e59581a267463b3efb671a748343 - we are now only throwing errors if there exists a sled reservation, for any reason.

This should fit better against concurrent creation requests, and it also nicely doesn't care *why* we have a reservation (for instance start, migration, etc) just that we have one.